### PR TITLE
Refactor Loop and Filter. Rename SubBerekening to Loop, add some extr…

### PR DESCRIPTION
**Pull request by vzorge**
…a convenience for Filter and clarify the Input and Output types of SubRunData

### CODE REVIEW

File: engine/src/main/scala/org/scalarules/dsl/nl/ScalaRulesDslTrait.scala [hunk 1]
```diff
@@ -1,16 +1,19 @@
...
-import org.scalarules.dsl.nl.grammar.{DslCondition, DslConditionImplicits, DslEvaluationImplicits, GegevenWord}
+import org.scalarules.dsl.nl.grammar.{DslCondition, DslConditionImplicits, DslEvaluationImplicits, DslListFilterWord, DslLoopWordTrait, GegevenWord}
 import org.scalarules.engine._
 
 /**
   * Aggregates the keywords and implicit definitions of the Scala-Rules DSL. The implicits available in this
   * trait can be used by importing the `grammar` package object's members, or extending this trait.
   */
-trait ScalaRulesDsl extends DslConditionImplicits
+trait ScalaRulesDslTrait extends DslConditionImplicits
```
Comment (jhkuperus) [inline engine/src/main/scala/org/scalarules/dsl/nl/ScalaRulesDslTrait.scala:11]: Why did you add the term `Trait` to the name?

Compare these two usages: 

``` scala
package object MyDsl extends ScalaRulesDsl

package object MyDsl extends ScalaRulesDslTrait
```

Long story short: I prefer the first one :)

(And the 80s called, they want their `nameString` and `ageInt` back ^_^)

File: engine/src/main/scala/org/scalarules/dsl/nl/grammar/DslLoop.scala [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  def over[A](listFact: ListFact[A]): DslLoopOverWord[A] = new DslLoopOverWord(listFact)
+}
+
+class DslLoopOverWord[A](listFact: ListFact[A]) {
+  def per(fact: SingularFact[A]): DslLoop[A] = new DslLoop(listFact, fact)
+}
+
+class DslLoop[A](listFact: ListFact[A], fact: SingularFact[A]) {
+  def doe[B, C](dslEvaluation: SubRunData[B, C]): SubRunData[List[B], A] = {
+    val resultFact: Fact[List[B]] = new ListFact("LoopFact")
```
Comment (jhkuperus) [inline engine/src/main/scala/org/scalarules/dsl/nl/grammar/DslLoop.scala:19]: How is this going to hold up when we are nesting loops? I remember trying this just before `SubBerekening` was introduced and not having a solution for this.

File: engine/src/main/scala/org/scalarules/dsl/nl/ScalaRulesDslTrait.scala (hunk 1 above)
Comment (vzorge) [inline engine/src/main/scala/org/scalarules/dsl/nl/ScalaRulesDslTrait.scala:11]: I don't know. this must have snuck in there by accident. Your way seems better for sure

File: engine/src/main/scala/org/scalarules/dsl/nl/grammar/DslLoop.scala (hunk 2 above)
Comment (vzorge) [inline engine/src/main/scala/org/scalarules/dsl/nl/grammar/DslLoop.scala:19]: We could randomize the LoopFact name. that would do it.
